protocol/contracts/governance/Templar.sol

```diff
@@ -0,0 +1,89 @@
+//SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+/**
+ * @title an NFT representing membership in the temple community.
+ *
+ * It exists as on chain proof that an account owns a discord id.
+ */
+contract Templar is ERC721, AccessControl {
+
+    bytes32 public constant OWNER = keccak256("OWNER");
+    bytes32 public constant CAN_ASSIGN = keccak256("CAN_ASSIGN");
+
+    string public baseUri = "https://discordapp.com/users/";
+
+     // Mapping from discordId ID to max temple role
+    mapping(uint256 => string) public templeRole;
+
+    constructor() ERC721("Templar", "TEMPLAR") {
+        _grantRole(OWNER, msg.sender);
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function supportsInterface(bytes4 interfaceId) public view override(ERC721, AccessControl) returns (bool) {
+        return (
+            ERC721.supportsInterface(interfaceId) || 
+            AccessControl.supportsInterface(interfaceId)
+        );
+    }
+
+    /**
+     *  Update the base URI for tokens.
+     */
+    function setBaseUri(string calldata _baseUri) external onlyRole(OWNER) {
+        baseUri = _baseUri;
+        emit BaseUriUpdated(baseUri);
+    }
+
+    /**
+     * Assign the NFT for the specified discord id delivering it to the specified
+     * address. 
+     * 
+     * This will either mint the NFT (if it doesn't exist), or transfer it.
+     *
+     * This method will be called by the discord bot when it is DM'd with a
+     * request of the form:
+                         /!claim_gov_nft <ADDR>
+     */
+    function assign( 
+        address account,
+        uint256 discordId,
+        string calldata _templeRole
+    ) external onlyRole(CAN_ASSIGN) {
+        if (account == address(0)) revert InvalidAddress(account);
+
+        if (_exists(discordId)) {
+            if (ownerOf(discordId) != account) {
+                _transfer(ownerOf(discordId), account, discordId);
+            }
+        } else {
+            _safeMint(account, discordId);
+        }
+
```

Comment:
Would add an event here e.g.:
```
Assigned(ownerOf(discordId), account, discordId, _templeRole)
```

Reply:
cleaner to leave mint/transfer since they emit Transfer already, and UpdateTempleRole is emitted separately?

Reply:
Makes it much harder to get all the needed info - we have to listen to 3 separate events instead of just 1.

Reply:
I didn't realise that relying on multiple events would make the subgraph implementation harder.

Is it normal to add new events to contracts in order to met subgraph needs?

Reply:
Why not? After all, `assign()` is not the same as `transfer()/mint()`. If the event provides value, it should be added. I would even argue if the correct way is to have events for transfer and mint. Do they provide useful information for our use case? We don't actually care if the token was transferred or minted. Always add events for the sepcific use case.

Reply:
> We don't actually care if the token was transferred or mint

We might eventually. It could well be useful to know when a discord user first went on chain (ie mint), and when a discord user changed their address (ie transfer)?

Reply:
I don't have experience with how much flexibility thegraph supports, but I assume it would be straightforward to produce 2 subgraphs

- one mapping address to discord id   (using Mint and Transfer events)
- the other mapping discord id to temple role (using the UpdateTempleRole event)

the election results calculated could use these as inputs directly.

Reply:
Updating the roles and assigning the NFTS are now seperate methods on 2 different contracts. So I'm, resolving this, as we can start a new thread if required.